src/binary_heap.rs

```diff
@@ -1086,6 +1097,47 @@ impl<T, C: Compare<T>> BinaryHeap<T, C> {
 
         self.rebuild_tail(start);
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` for which `f(&e)` returns
+    /// `false`. The elements are visited in unsorted (and unspecified) order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use std::collections::BinaryHeap;
+    ///
+    /// let mut heap = BinaryHeap::from([-10, -5, 1, 2, 4, 13]);
+    ///
+    /// heap.retain(|x| x % 2 == 0); // only keep even numbers
+    ///
+    /// assert_eq!(heap.into_sorted_vec(), [-10, 2, 4])
+    /// ```
+    // #[stable(feature = "binary_heap_retain", since = "1.70.0")]
+    pub fn retain<F>(&mut self, mut f: F)
```

Comment:
We could have an optimized `remove_one(&mut self, find_fn: FnMut(&T) -> bool)`  that only removes the first matching element with [`Vec::swap_remove` ](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.swap_remove) before rebuilding the heap from that location (same `RebuildOnDrop`). This removes half the comparisons in average and the memory move.

Reply:
that's a brilliant idea, I have added `remove_once` based on your suggestion. We probably don't need `retain` anymore if we have that.